pkg/machinery/client/connection.go

```diff
@@ -0,0 +1,183 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+
+	clientconfig "github.com/talos-systems/talos/pkg/machinery/client/config"
+	"github.com/talos-systems/talos/pkg/machinery/client/resolver"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+	"github.com/talos-systems/talos/pkg/machinery/generic/slices"
+)
+
+// getConn creates new gRPC connection.
+func (c *Client) getConn(ctx context.Context, opts ...grpc.DialOption) (*GRPCConnectionWrapper, error) {
+	endpoints := c.GetEndpoints()
+
+	target := c.getTarget(
+		resolver.EnsureEndpointsHavePorts(
+			reduceURLsToAddresses(endpoints),
+			constants.ApidPort),
+	)
+
+	dialOpts := []grpc.DialOption(nil)
+
+	dialOpts = append(dialOpts, c.options.grpcDialOptions...)
+
+	dialOpts = append(dialOpts, opts...)
+
+	if c.options.unixSocketPath != "" {
+		conn, err := grpc.DialContext(ctx, target, dialOpts...)
+
+		return NewGRPCConnectionWrapper(c.GetConfigContext().Cluster, conn), err
+	}
+
+	tlsConfig := c.options.tlsConfig
+
+	if tlsConfig != nil {
+		return c.makeConnection(ctx, target, credentials.NewTLS(tlsConfig), dialOpts)
+	}
+
+	tlsConfig = &tls.Config{}
+
+	if err := c.resolveConfigContext(); err != nil {
+		return nil, fmt.Errorf("failed to resolve configuration context: %w", err)
+	}
+
+	basicAuth := c.options.configContext.Auth.Basic
+	if basicAuth != nil {
+		dialOpts = append(dialOpts, WithGRPCBasicAuth(basicAuth.Username, basicAuth.Password))
+		if shouldInsecureConnectionsBeAllowed(endpoints) {
+			return c.makeConnection(ctx, target, insecure.NewCredentials(), dialOpts)
+		}
+	}
+
+	caBytes, err := getCA(c.options.configContext)
```

Comment:
lines 68-91 here seems to be building `credentials.TransportCredentials` and we can extract this to a separate (testable) function. 

in this function we can also return `Insecure` creds if it's a debug build and insecure creds were requested